query/join_test.go

```diff
@@ -0,0 +1,127 @@
+
+	emptyList = api.SeriesList{[]api.Timeseries{}, api.Timerange{}}
+	basicList = api.SeriesList{[]api.Timeseries{seriesA1, seriesA2, seriesB3, seriesB4, seriesC5}, api.Timerange{}}
+	dcList    = api.SeriesList{[]api.Timeseries{seriesDC_A, seriesDC_B, seriesDC_C}, api.Timerange{}}
+	envList   = api.SeriesList{[]api.Timeseries{seriesENV_PROD, seriesENV_STAGE}, api.Timerange{}}
+
+	voidList = api.SeriesList{[]api.Timeseries{voidSeries}, api.Timerange{}}
+)
+
+func Test_Join_Zero(t *testing.T) {
+	// attempt to join empty with empty, empty with many, and many with empty, etc.
+	// all should produce empty results
+	result := Join([]api.SeriesList{emptyList})
+	if len(result.Rows) != 0 {
+		t.Fatalf("Join of single empty row produces non-empty result")
+	}
+	result = Join([]api.SeriesList{emptyList, emptyList})
+	if len(result.Rows) != 0 {
+		t.Fatalf("Join of two empty rows produces non-empty result")
+	}
+	result = Join([]api.SeriesList{emptyList, basicList})
+	if len(result.Rows) != 0 {
+		t.Fatalf("Join of (empty row, nonempty row) produces non-empty result")
+	}
+	result = Join([]api.SeriesList{basicList, emptyList})
+	if len(result.Rows) != 0 {
+		t.Fatalf("Join of single (nonempty row, empty row) produces non-empty result")
+	}
+	result = Join([]api.SeriesList{basicList, basicList, basicList, emptyList, basicList})
+	if len(result.Rows) != 0 {
+		t.Fatalf("Join of many nonempty rows and one empty row produces non-empty result")
+	}
+}
+
+func Test_Join_Self(t *testing.T) {
+	// attempt to join (well-behaved) sets with themselves
+	// or with nothing
+	result := Join([]api.SeriesList{basicList})
+	if len(result.Rows) != len(basicList.Series) {
+		t.Fatalf("Join of a row with nothing else changes its length")
+	}
+
+	result = Join([]api.SeriesList{basicList, basicList})
+	if len(result.Rows) != len(basicList.Series) {
+		t.Fatalf("Join of a row with itself changes its length")
+	}
+
+	result = Join([]api.SeriesList{dcList, dcList})
+	if len(result.Rows) != len(dcList.Series) {
+		t.Fatalf("Join of a row with itself changes its length")
+	}
+
+	result = Join([]api.SeriesList{envList, envList})
+	if len(result.Rows) != len(envList.Series) {
+		t.Fatalf("Join of a row with itself changes its length")
+	}
+}
+
+func Test_Join_CartesianProduct(t *testing.T) {
+	// attempt to join two different sets with nothing in common
+	// and check that they multiply
+	result := Join([]api.SeriesList{basicList, envList})
+	if len(result.Rows) != len(basicList.Series)*len(envList.Series) {
+		t.Fatalf("Join with serieslists having disjoint key sets results in unexpected number of entries (should be %v but is %v)", len(basicList.Series)*len(envList.Series), len(result.Rows))
+	}
+}
+
+// A simple max function used in the Test_Join_KeySubsets which makes it less brittle
+// to modifications to the series lists used as example testcases
+func max(x, y int) int {
+	if x < y {
+		return y
+	} else {
+		return x
+	}
+}
+
+func Test_Join_KeySubsets(t *testing.T) {
```

Comment:
i think i might combine both tests (this `Join_KeySubsets` and `Join_Single`), and do table-based testing. i.e. store the test suites (input + output pair) in an array, and simply iterate through them. you can look at files in `query_test.go` or other files for more info.